text/0517-io-os-reform.md

```diff
@@ -708,7 +712,129 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::fs`
 [std::fs]: #stdfs
 
-> To be added in a follow-up PR.
+The `fs` module will provide most of the functionality it does today,
+but with a stronger cross-platform orientation.
+
+Note that all path-consuming functions will now take an
+`AsPath`-bounded parameter for ergonomic reasons (this will allow
+passing in Rust strings and literals directly, for example).
+
+#### Free functions
+[Free functions]: #free-functions
+
+**Files**:
+
+* `copy`. Take `AsPath` bound.
+* `rename`. Take `AsPath` bound.
+* `remove_file` (renamed from `unlink`). Take `AsPath` bound.
+
+* `file_attr` (renamed from `stat`). Take `AsPath` bound. Yield a new
+  struct, `FileAttr`, with no public fields, but `size`, `kind` and
+  `perm` accessors. The various `os::platform` modules will offer
+  extension methods on this structure.
+
+* `set_perm` (renamed from `chmod`). Take `AsPath` bound, and a
+  `FilePermissions` value. The `FilePermissions` type will be revamped
+  as a struct with private implementation; see below.
+
+**Directories**:
+
+* `make_dir` (renamed from `mkdir`). Take `AsPath` bound.
+* `make_dir_all` (renamed from `mkdir_recursive`). Take `AsPath` bound.
+* `read_dir` (renamed from `readdir`). Take `AsPath` bound. Yield a
+  newtypes iterator, which yields a new type `DirEntry` which has an
+  accessor for `Path`, but will eventually provide other information
+  as well (possibly via platform-specific extensions).
+* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
+* `remove_dir_all` (renamed from `rmdir_recursive`). Take
+  `AsPath` bound.
+* `walk_dir`. Take `AsPath` bound. Yield an iterator over `IoResult<DirEntry>`.
+
+**Links**:
+
+* `hard_link` (renamed from `link`). Take `AsPath` bound.
+* `sym_link` (renamed from `symlink`). Take `AsPath` bound.
+* `read_link` (renamed form `readlink`). Take `AsPath` bound.
+
+#### Files
+[Files]: #files
+
+The `File` type will largely stay as it is today, except that it will
+use the `AsPath` bound everywhere.
```

Comment:
On Wed, Jan 28, 2015 at 10:39 AM, Steven Fackler <notifications@github.com>
wrote:
>
> In what way is that unsafe? The fd is just an integer, and I can't see how
> you'd extract any memory unsafety from it. In the worst case, it's closed
> while there's another copy of it floating around, in which case calls using
> it start returning errors.
>
No, the worst case is it *aliases another later opened file* and you
truncate /etc/passwd without meaning to.  You can also pass it to open
/proc/self/fd/N and get the wrong (aliased) file.   Heh.

>